From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Etaash Mathamsetty
 <45927311+Etaash-mathamsetty@users.noreply.github.com>
Date: Fri, 10 Apr 2026 00:14:01 -0500
Subject: [PATCH 15/46] 
 0020-winewayland-Implement-app-triggered-window-restore.patch

---
 dlls/winewayland.drv/window.c | 15 +++++++++++++++
 1 file changed, 15 insertions(+)

diff --git a/dlls/winewayland.drv/window.c b/dlls/winewayland.drv/window.c
index 64461729def..aa7ce857a67 100644
--- a/dlls/winewayland.drv/window.c
+++ b/dlls/winewayland.drv/window.c
@@ -201,6 +201,7 @@ static BOOL wayland_win_data_create_wayland_surface(struct wayland_win_data *dat
     struct wayland_client_surface *client = data->client_surface;
     struct wayland_surface *surface;
     enum wayland_surface_role role;
+    HWND focused;
     BOOL visible, server_decor = FALSE;
     DWORD exstyle = NtUserGetWindowLongW(data->hwnd, GWL_EXSTYLE);
     DWORD style = NtUserGetWindowLongW(data->hwnd, GWL_STYLE);
@@ -239,6 +240,20 @@ static BOOL wayland_win_data_create_wayland_surface(struct wayland_win_data *dat
         server_decor = TRUE;
     }
 
+    pthread_mutex_lock(&process_wayland.keyboard.mutex);
+    focused = process_wayland.keyboard.focused_hwnd;
+    pthread_mutex_unlock(&process_wayland.keyboard.mutex);
+
+    /* Force recreate the toplevel if it is now unminimized.
+     * This resets the toplevel state, effectively making it unminimized.
+     * Ensure that this restoration is not user driven,
+     * as this case is already handled by the compositor */
+    if (surface->window.minimized && focused != data->hwnd
+        && !(style & WS_MINIMIZE))
+    {
+        wayland_surface_clear_role(surface);
+    }
+
     /* If the window is a visible toplevel make it a wayland
      * xdg_toplevel. Otherwise keep it role-less to avoid polluting the
      * compositor with empty xdg_toplevels. */
